### PR TITLE
Omnibus PR containing multiple improvements

### LINKED ISSUE #15: is `StructTypes.isempty` part of the public API?
I'd like to pirate
```julia
StructTypes.isempty(::Missing) = true
```
but it would be good to know if that's an internal method (and so I should pin the version instead of relying on semver). I suspect the answer is yes by default since it's not documented or exported, so I guess this is also a request that it would be nice to be able to avoid writing fields (with JSON3) which are missing if they show up in `StructTypes.omitempties`.

### LINKED ISSUE #22: Bug: when calling `applyfield!` on a mutable struct, the 32nd field is skipped
For example, suppose that you have a mutable struct with 40 fields. And you call `applyfield!` on an instance of this mutable struct. Fields 1 to 31 will be applied correctly. And fields 33 to 40 will be applied correctly. But field number 32 will always be skipped.

This is a big problem, as it means the 32nd field in the struct will always be skipped when e.g. reading JSON in to the struct.

I discovered this bug for the `applyfield!` function, but it is possible that the bug also exists in the other functions.

Full example:
```julia
julia> import StructTypes

julia> mutable struct Foo
           field_1
           field_2
           field_3
           field_4
           field_5
           field_6
           field_7
           field_8
           field_9
           field_10
           field_11
           field_12
           field_13
           field_14
           field_15
           field_16
           field_17
           field_18
           field_19
           field_20
           field_21
           field_22
           field_23
           field_24
           field_25
           field_26
           field_27
           field_28
           field_29
           field_30
           field_31
           field_32
           field_33
           field_34
           field_35
           field_36
           field_37
           field_38
           field_39
           field_40
           Foo() = new()
       end

julia> length(fieldnames(Foo))
40

julia> for fieldname in fieldnames(Foo)
           f_applied = StructTypes.applyfield!((x, y, z) -> nothing, Foo(), fieldname)
           @info "" fieldname f_applied
       end
┌ Info:
│   fieldname = :field_1
└   f_applied = true
┌ Info:
│   fieldname = :field_2
└   f_applied = true
┌ Info:
│   fieldname = :field_3
└   f_applied = true
┌ Info:
│   fieldname = :field_4
└   f_applied = true
┌ Info:
│   fieldname = :field_5
└   f_applied = true
┌ Info:
│   fieldname = :field_6
└   f_applied = true
┌ Info:
│   fieldname = :field_7
└   f_applied = true
┌ Info:
│   fieldname = :field_8
└   f_applied = true
┌ Info:
│   fieldname = :field_9
└   f_applied = true
┌ Info:
│   fieldname = :field_10
└   f_applied = true
┌ Info:
│   fieldname = :field_11
└   f_applied = true
┌ Info:
│   fieldname = :field_12
└   f_applied = true
┌ Info:
│   fieldname = :field_13
└   f_applied = true
┌ Info:
│   fieldname = :field_14
└   f_applied = true
┌ Info:
│   fieldname = :field_15
└   f_applied = true
┌ Info:
│   fieldname = :field_16
└   f_applied = true
┌ Info:
│   fieldname = :field_17
└   f_applied = true
┌ Info:
│   fieldname = :field_18
└   f_applied = true
┌ Info:
│   fieldname = :field_19
└   f_applied = true
┌ Info:
│   fieldname = :field_20
└   f_applied = true
┌ Info:
│   fieldname = :field_21
└   f_applied = true
┌ Info:
│   fieldname = :field_22
└   f_applied = true
┌ Info:
│   fieldname = :field_23
└   f_applied = true
┌ Info:
│   fieldname = :field_24
└   f_applied = true
┌ Info:
│   fieldname = :field_25
└   f_applied = true
┌ Info:
│   fieldname = :field_26
└   f_applied = true
┌ Info:
│   fieldname = :field_27
└   f_applied = true
┌ Info:
│   fieldname = :field_28
└   f_applied = true
┌ Info:
│   fieldname = :field_29
└   f_applied = true
┌ Info:
│   fieldname = :field_30
└   f_applied = true
┌ Info:
│   fieldname = :field_31
└   f_applied = true
┌ Info:
│   fieldname = :field_32
└   f_applied = false
┌ Info:
│   fieldname = :field_33
└   f_applied = true
┌ Info:
│   fieldname = :field_34
└   f_applied = true
┌ Info:
│   fieldname = :field_35
└   f_applied = true
┌ Info:
│   fieldname = :field_36
└   f_applied = true
┌ Info:
│   fieldname = :field_37
└   f_applied = true
┌ Info:
│   fieldname = :field_38
└   f_applied = true
┌ Info:
│   fieldname = :field_39
└   f_applied = true
┌ Info:
│   fieldname = :field_40
└   f_applied = true
```

### PR DESCRIPTION
**Pull request by DilumAluthge**
This PR makes the following changes:
1. Bug fix: fixes a bug in which `applyfield!` would never be applied to the 32nd field of a mutable struct (#22) 
2. New feature: adds the public `makeobj` and `makeobj!` functions
3. New feature: adds the ability to automatically omit all empty fields in a struct 
4. New feature: adds the ability for users to override the behavior of `isempty` for their own structs 
5. Test suite: makes some small improvements to a few existing unit tests
6. Test suite: Sets up GitHub Actions CI on this repository
7. Bumps the version number in `Project.toml` from `1.1.0` to `1.2.0`

### Fixes the following issues:

Fixes quinnj/JSON3.jl#78 
Fixes quinnj/JSON3.jl#90 
Fixes #15 
Fixes #22 

### Closes the following pull requests:

Closes #16 
Closes #19 
Closes #20 
Closes #21

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    tags: '*'
+jobs:
+  test:
+    name: Julia ${{ matrix.version }} - ${{ matrix.os }} - ${{ matrix.arch }} - ${{ github.event_name }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        version:
+          - '1.2'
```
Comment (quinnj) [inline .github/workflows/ci.yml:18]: Why 1.2 instead of 1.0?

File: Project.toml [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
+version = "1.2.0"
 
 [deps]
-UUIDs = "cf7118a7-6976-5b1a-9a39-7adc72f591a4"
 Dates = "ade2ca70-3891-5945-98fb-dc099432e06a"
+UUIDs = "cf7118a7-6976-5b1a-9a39-7adc72f591a4"
 
 [compat]
-julia = "1"
+julia = "1.2"
```
Comment (quinnj) [inline Project.toml:11]: Why this change?

File: src/StructTypes.jl [hunk 3]
```diff
@@ -531,7 +543,12 @@ Note that any `StructTypes.names` mappings are applied, as well as field-specifi
              x_17, x_18, x_19, x_20, x_21, x_22, x_23, x_24, x_25, x_26, x_27, x_28, x_29, x_30, x_31, x_32, vals...)
 end
 
-Base.@pure function symbolin(names::Tuple{Vararg{Symbol}}, name::Symbol)
+Base.@pure function symbolin(names::Union{Tuple{Vararg{Symbol}}, Bool}, name::Symbol)
+    if names === true
```
Comment (quinnj) [inline src/StructTypes.jl:547]:
```suggestion
    names isa Bool && return names
```

File: src/StructTypes.jl [hunk 4]
```diff
@@ -132,6 +142,8 @@ isempty(::Number) = false
 isempty(::Nothing) = true
 isempty(x) = false
 isempty(x, i) = isempty(Core.getfield(x, i))
+isempty(::Type{T}, x) where {T} = isempty(x) # generic fallback
```
Comment (quinnj) [inline src/StructTypes.jl:145]: Can you walk through why this is needed? What's your use-case?

File: src/StructTypes.jl [hunk 5]
```diff
@@ -656,7 +673,7 @@ mappings will be applied, and the function will be passed the Julia field name.
     f_applied = false
     # unroll the first 32 field checks to avoid dynamic dispatch if possible
     Base.@nif(
-        32,
+        33,
```
Comment (quinnj) [inline src/StructTypes.jl:676]: Why is this changed?

File: src/StructTypes.jl [hunk 6]
```diff
@@ -686,4 +703,57 @@ mappings will be applied, and the function will be passed the Julia field name.
     return f_applied
 end
 
+"""
+    makeobj(::Type{Any}, obj::Any)
```
Comment (quinnj) [inline src/StructTypes.jl:707]: I wonder if we just call this `construct`; we already have that defined in StructTypes, but that's really what we're doing here, and custom types are still able to overload it if they want.

File: src/StructTypes.jl [hunk 7]
```diff
@@ -686,4 +703,57 @@ mappings will be applied, and the function will be passed the Julia field name.
...
+    makeobj(::Type{T}, obj::AbstractDict) where {T <: AbstractDict}
+
+Convert `obj` to an object of `T`.
+"""
+makeobj(::Type{T}, obj::AbstractDict) where {T <: AbstractDict} = convert(T, obj)::T
+
+"""
+    makeobj(::Type{T}, obj::AbstractDict) where {T}
+
+Convert `obj` to an object of `T`. `T` must be a mutable struct.
```
Comment (quinnj) [inline src/StructTypes.jl:737]: `T` must be a `StructTypes.Mutable()`

File: src/StructTypes.jl (hunk 5 above)
Comment (quinnj) [reply src/StructTypes.jl:676]: Ok, saw your comment in the PR

**Review by quinnj — COMMENTED** [review]
Good overall; I prefer PRs to be much smaller, because most of this could be merged right away, but there are parts that need more discussion/iteration.

File: src/StructTypes.jl (hunk 5 above)
Comment (DilumAluthge) [reply src/StructTypes.jl:672]: Yeah, without this change, the 32nd field of a struct will never get applied.

File: src/StructTypes.jl (hunk 4 above)
Comment (DilumAluthge) [reply src/StructTypes.jl:145]: This allows users to customize the behavior of `isempty` for their own structs without modifying the behavior of `isempty` for other types or committing any type piracy.

For example, suppose that I want `missing` to be treated as empty when serializing my structs. I can just do `isempty(::Type{MyType}, ::Missing) = true`.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (DilumAluthge) [reply .github/workflows/ci.yml]: The `hasfield` function is only available on Julia 1.2 and higher. But, I figured out a way to make it work on Julia 1.0 and 1.1, so now we test on Julia 1.0 and up.

File: Project.toml (hunk 2 above)
Comment (DilumAluthge) [reply Project.toml]: Originally this was because of `hasfield` requiring Julia 1.2. But now we have a way of making it work on Julia 1.0 and 1.1, so this change no longer exists.

File: src/StructTypes.jl (hunk 7 above)
Comment (DilumAluthge) [reply src/StructTypes.jl]: Done.

File: src/StructTypes.jl (hunk 6 above)
Comment (DilumAluthge) [reply src/StructTypes.jl:703]: Hmmm, personally I'd find it confusing if this functionality was also named `construct`.

**Comment by DilumAluthge** [comment]
> I prefer PRs to be much smaller

I agree 100%. I almost never make PRs this big. The only reason I did so in this case is that when I tried merging the individual PRs locally, I realized that there were a lot of merge conflicts between the individual PRs.

File: src/StructTypes.jl (hunk 4 above)
Comment (DilumAluthge) [reply src/StructTypes.jl:145]: A little more explanation is here: https://github.com/JuliaData/StructTypes.jl/pull/21

**Comment by DilumAluthge** [comment]
@quinnj This is ready for a second round of review.

**Comment by quinnj** [comment]
Ok, merge this for now; but I'd like to work on the `makeobj` more; I'd like it to be more general and make sure we're using/respecting all the StructTypes machinery.
